kotlin/samples/recyclerview/src/main/java/com/squareup/sample/recyclerview/editablelistworkflow/EditableListAdapter.kt

```diff
@@ -0,0 +1,135 @@
+import com.squareup.sample.recyclerview.editablelistworkflow.EditableListWorkflow.Rendering
+import com.squareup.sample.recyclerview.editablelistworkflow.ListDiffMode.Asynchronous
+import com.squareup.sample.recyclerview.editablelistworkflow.ListDiffMode.None
+import com.squareup.sample.recyclerview.editablelistworkflow.ListDiffMode.Synchronous
+import com.squareup.sample.recyclerview.inputrows.viewType
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import timber.log.Timber
+
+class EditableListAdapter : Adapter<ViewHolder>() {
+
+  private val scope = CoroutineScope(Dispatchers.Main)
+  private var diffJob: Job = Job()
+  private var rendering: Rendering? = null
+  private val rows: List<RowValue<*, *>> get() = checkNotNull(rendering?.rowValues)
+
+  fun updateRendering(
+    newRendering: Rendering,
+    diffMode: ListDiffMode
+  ) {
+    diffJob.cancel()
+    Timber.i("Updating adapter list: $diffMode")
+    when (diffMode) {
+      None -> {
+        rendering = newRendering
+        notifyDataSetChanged()
+      }
+      Synchronous -> {
+        val diffCallback = DiffCallback(rendering, newRendering)
+        val diffResult = DiffUtil.calculateDiff(diffCallback, false)
+        rendering = newRendering
+        diffResult.dispatchUpdatesTo(this@EditableListAdapter)
+      }
+      Asynchronous -> {
+        val diffCallback = DiffCallback(rendering, newRendering)
+        diffJob = scope.launch {
+          val diffResult = withContext(Dispatchers.Default) {
+            Timber.i("Starting async diff")
+            // Fake delay to simulate long diff calculation.
+            delay(500)
+            DiffUtil.calculateDiff(diffCallback, false)
+          }
+          Timber.i("Async diff ready")
+          rendering = newRendering
+          diffResult.dispatchUpdatesTo(this@EditableListAdapter)
+        }
+      }
+    }
+  }
+
+  override fun getItemCount(): Int = rendering?.rowValues?.size ?: 0
+
+  /**
+   * Items are always in a fixed order.
+   */
+  override fun getItemId(position: Int): Long = position.toLong()
+
+  /**
+   * Every position is a different type.
+   */
+  override fun getItemViewType(position: Int): Int = rows[position].row.viewType
+
+  override fun onCreateViewHolder(
+    parent: ViewGroup,
+    viewType: Int
+  ): ViewHolder {
+    Timber.i("Creating new ViewHolder for type $viewType…")
+    return rows.first { it.row.viewType == viewType }
+        .row.createViewHolder(parent)
+  }
+
+  override fun onBindViewHolder(
+    holder: ViewHolder,
+    position: Int
+  ) {
+    Timber.i("Binding $position to $holder…")
+    rendering?.let { rendering ->
+      val rowValue = rendering.rowValues[position]
+      // Disable the listener first, so we don't try sending view events while we're updating
```

Comment:
should be scope/extension function on `rowValue` instead to encapsulate listener trickery